datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -816,14 +831,14 @@ fn sort_batch(
     batch: RecordBatch,
     schema: SchemaRef,
     expr: &[PhysicalSortExpr],
+    fetch: Option<usize>,
 ) -> ArrowResult<BatchWithSortArray> {
-    // TODO: pushup the limit expression to sort
     let sort_columns = expr
         .iter()
         .map(|e| e.evaluate_to_sort_column(&batch))
         .collect::<Result<Vec<SortColumn>>>()?;
 
-    let indices = lexsort_to_indices(&sort_columns, None)?;
+    let indices = lexsort_to_indices(&sort_columns, fetch)?;
```

Comment:
nice

Reply:
I wonder if this will effectively get us much of the benefit of a special TopK operator as we don't have to copy the entire input -- we only copy the `fetch` limit, if specified

Although I suppose `SortExec` still buffers all of its input where a TopK could buffer them

Reply:
In fact, I wonder if you could also apply the limit here:

https://github.com/apache/arrow-datafusion/blob/3a9e0d0/datafusion/core/src/physical_plan/sorts/sort.rs#L123-L124

as part of sorting each batch -- rather than keeping the entire input batch, we only need to keep at most `fetch` rows from each batch.

Reply:
`lexsort_to_indices` already returns only `fetch` indices per batch, this is used to `take` that nr. of indices per batch, throwing away the rest of the rows.

 The remaining optimization I think is tweaking `SortPreservingMergeStream` to only maintain `fetch` records in the heap instead of all `fetch` top records for each batch in the partition as mentioned here https://github.com/apache/arrow-datafusion/issues/3516#issuecomment-1250807415. After this I think we have a full TopK implementation that only needs to keep  n number of rows in memory (per partition).

I would like to do this in a separate PR.

Reply:
A separate PR is a great idea 👍 

> lexsort_to_indices already returns only fetch indices per batch, this is used to take that nr. of indices per batch, throwing away the rest of the rows.

Right, the point I was trying to make is that there are 2 calls to `lexsort_to_indices` in sort.rs. I think this PR only pushed `fetch` to one of them. The second is https://github.com/apache/arrow-datafusion/blob/3a9e0d0/datafusion/core/src/physical_plan/sorts/sort.rs#L826 and I think it is correct to push `fetch` there too

I was thinking if we applied `fetch` to the second call, we could get close to the same effect without changing `SortPreservingMergeStream`. 

* After this PR, sort buffers `num_input_batches * input_batch_size` rows. 
* Adding `fetch` to the other call to `lexsort_to_indices` would would buffer `num_input_batches * limit` rows
* Extending `SortPreservingMergeStream` would allow us to buffer only `limit` rows.

So clearly extending `SortPreservingMergeStream` is optimal in terms of rows buffered, but it likely requires a bit more effort.

Reply:
Ah, I didn't look to much at the rest of the implementation, I think you're right that providing `limit` to the other `lexsort_to_indices` would be beneficial as well. I will create a issue for this and issue a PR later.

Reply:
I think the current change already buffers `num_input_batches * limit` by the way, as it is applied before adding them to the buffer. As far as I can see adding the second to `lexsort_to_indices` will reduce mainly the output of the individual sorts to `fetch` - which is of course beneficial too as that reduces time to sort and limit the input again to `take` and input to `SortPreservingMergeStream`

Reply:
> I think you're right that providing fetch to the other lexsort_to_indices would be beneficial as well. I will create a issue for this and issue a PR later.

for other readers, this is addressed by https://github.com/apache/arrow-datafusion/issues/3544 and fixed by https://github.com/apache/arrow-datafusion/pull/3545